### PR TITLE
Initial Gubbins recipe

### PR DESCRIPTION
**Pull request by andrewjpage**
This is a recipe for Gubbins for detecting recombination in bacteria. http://nar.oxfordjournals.org/content/43/3/e15

I've tested the recipe on:
OSX 10.9
OSX 10.10
Ubuntu 14.04
Ubunut 12.04
CentOS 7
CentOS 6.6

and theres detailed installation instructions available from here:
https://github.com/sanger-pathogens/gubbins/blob/master/INSTALL.md

Please let me know if it can be improved/simplified in any way.

### CODE REVIEW

**Comment by tdsmith** [comment]
@BrewTestBot test this please

**Comment by tdsmith** [comment]
Try running `brew audit --strict ./gubbins.rb`; this looks like a solid contribution but there will be several style notes.

File: gubbins.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  # doi "10.1093/nar/gku1196"
+
+  homepage 'https://github.com/sanger-pathogens/gubbins'
+  url 'https://github.com/sanger-pathogens/gubbins/archive/v1.4.0.tar.gz'
+  sha1 'adf2d25eb1740155370f3c33c07c1622d4703c0c'
+  head 'https://github.com/sanger-pathogens/gubbins.git'
+
+  depends_on :autoconf
+  depends_on :automake
+  depends_on :libtool
```
Comment (tdsmith) [inline gubbins.rb]: These are deprecated in favor of e.g. `depends_on "autoconf" => :build`.

File: gubbins.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  depends_on 'fasttree'
+  depends_on 'fastml' => :recommended
+  depends_on :python3
+  depends_on 'zlib'
+  
+  def install
+    inreplace "src/Makefile.am", "-lrt", "" if OS.mac? # no librt for osx
+    inreplace "configure.ac", "PKG_CHECK_MODULES([zlib], [zlib])", "AC_CHECK_LIB(zlib, zlib)" if OS.mac?
+    
+    system "cat python/Makefile.am"
```
Comment (tdsmith) [inline gubbins.rb]: This line can go

File: gubbins.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+           "--prefix=#{prefix}"
+    system "make","check"
+    system "make","install"
+  end
+
+  test do
+    system "gubbins"
+    system "run_gubbins.py"
+    system "gubbins_drawer.py"
+  end
```
Comment (tdsmith) [inline gubbins.rb]: Is it possible to feed gubbins some trivial input?

File: gubbins.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  depends_on :autoconf
+  depends_on :automake
+  depends_on :libtool
+  depends_on 'check'
+  depends_on 'raxml'
+  depends_on 'fasttree'
+  depends_on 'fastml' => :recommended
+  depends_on :python3
+  depends_on 'zlib'
```
Comment (tdsmith) [inline gubbins.rb]: `zlib` should be `homebrew/dupes/zlib` and the system version should be used on OS X unless it's too old

(so: `depends_on "homebrew/dupes/zlib" unless OS.mac?`, probably)

File: gubbins.rb (hunk 1 above)
Comment (tdsmith) [inline gubbins.rb]: And OS X provides libtool, so the libtool dependency should be conditional on OS.

**Comment by andrewjpage** [comment]
Thanks for the feedback.
I've updated the recipe but have only been able to check it works on Linux. I wont have access to OSX yosemite/mavericks until next week. It would be great if someone could double check it still works on OSX

File: gubbins.rb (hunk 3 above)
Comment (tseemann) [inline gubbins.rb]: Can you put both the OSX specific mods in a common  `if OS.mac? end` block?

Comment (tseemann) [inline gubbins.rb]: tag and doi need to come after the desc and homepage.

Comment (tseemann) [inline gubbins.rb]: if you have `head` can you please test it worked with `build gubbins --HEAD` (not sure of exact syntax)

Comment (tseemann) [inline gubbins.rb]: Are there any non-standard Python3 modules it depends on?

**Comment by tdsmith** [comment]
@BrewTestBot test this please

**Comment by tdsmith** [comment]
Build error:
```
==> autoreconf -i
configure.ac:10: error: possibly undefined macro: AC_PROG_LIBTOOL
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/local/Cellar/autoconf/2.69/bin/autoconf failed with exit status: 1
```

Audit warnings:
```
==> brew style gubbins
== /usr/local/Library/Taps/homebrew/homebrew-science/gubbins.rb ==
C:  5: 12: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C:  6:  7: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C:  7: 10: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C:  8:  8: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 13: 14: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 14: 14: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 15: 14: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 16: 14: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 18: 14: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 24: 24: Space missing after comma.
C: 29: 18: Space missing after comma.
C: 30: 18: Space missing after comma.

1 file inspected, 12 offenses detected
```

**Comment by tdsmith** [comment]
This may need to use a brewed libtool after all.

**Comment by andrewjpage** [comment]
Thanks a million for the help.
I've updated it so that it passes the audit and style checks, fixing spaces, quotes, order of lines, and grouped mac specific stuff together.
It has non standard python3 dependancies so I've added a check for those (they were installed automatically which I believe was a bit of a no no), which prompts for you to use pip3 to install (and updated my install docs to use pip before running brew).
Added back libtool and checked it works with --HEAD. 
I've only checked it on Linux because I dont have access to an OSX dev system for a few days.

@BrewTestBot test this please

**Comment by tseemann** [comment]
@andrewjpage @tdsmith 

I've pulled it in. Now note I do **not** have python3 on my macbook air:

Shouldn't it complain about `python3` first, before it complains about python3 modules?

```
brew reinstall gubbins
==> Reinstalling homebrew/science/gubbins
gubbins: Unsatisfied dependency: biopython
Homebrew does not provide Python3 dependencies; install with:
  pip3 install biopython
gubbins: Unsatisfied dependency: nose
Homebrew does not provide Python3 dependencies; install with:
  pip3 install nose
gubbins: Unsatisfied dependency: dendropy
Homebrew does not provide Python3 dependencies; install with:
  pip3 install dendropy
gubbins: Unsatisfied dependency: reportlab
Homebrew does not provide Python3 dependencies; install with:
  pip3 install reportlab
```

**Comment by tseemann** [comment]
Do you have issues with the FastTree single precision rounding problem?

if so, I think you specify this:
```
depends_on "fasttree" => ["with-double", :recommended]
```

```
% brew info fasttree
--with-double
        Use double precision maths for accurate branch lengths (disables SSE)
```

**Comment by tseemann** [comment]
For me the test failed:

```
brew test gubbins
Testing homebrew/science/gubbins
==> gubbins
  -a    Min window size
  -b    Max window size
  -h    Display this usage information.

Error: File '' doesnt exist
Error: homebrew/science/gubbins: failed
Failed executing: gubbins
/usr/local/Library/Homebrew/formula.rb:897:in `block in system'
/usr/local/Library/Homebrew/formula.rb:854:in `open'
/usr/local/Library/Homebrew/formula.rb:854:in `system'
/usr/local/Library/Taps/homebrew/homebrew-science/gubbins.rb:43:in `block in <class:Gubbins>'
/usr/local/Library/Homebrew/formula.rb:797:in `block in run_test'
/usr/local/Library/Homebrew/extend/fileutils.rb:17:in `mktemp'
/usr/local/Library/Homebrew/formula.rb:793:in `run_test'
/usr/local/Library/Homebrew/test.rb:28:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Library/Homebrew/test.rb:27:in `<main>'
```

**Comment by tseemann** [comment]
I'm getting style issues too:
```
brew style gubbins
== gubbins.rb ==
C: 25:  1: Extra blank line detected.
C: 27:  1: Extra empty line detected at method body beginning.
1 file inspected, 2 offenses detected
```

**Comment by tseemann** [comment]
@andrewjpage Can you squash all your commits into one?  Maybe that's my issue.

**Comment by tseemann** [comment]
I note that `gubbins` is not meant to be run directly, but rather via `run_gubbins.py`. Usually these "helper exes" should be installed via `libexec.install` and the master script patched to run it from there. This avoids non-user exes being in the `PATH`.

As an aside, I would rename `run_gubbins.py` to `gubbins` and make the binary the longer name (and in `libexec`).

**Comment by tdsmith** [comment]
Python module dependencies should be installed as `resource`s. You can see jrnl.rb in core for an example.

**Comment by tdsmith** [comment]
There's some documentation on how we like to handle Python module dependencies here: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Python-for-Formula-Authors.md#dependencies

**Comment by tseemann** [comment]
Except we still have the Linux `lib64` python bug for that method: https://github.com/Homebrew/linuxbrew/issues/443

Gubbins will primarily be used on Linux servers due to computational needs.

**Comment by tdsmith** [comment]
That's a quirk of certain Pythons which I think Linuxbrew should not attempt to support. I think the resolution is to document it in Linuxbrew. I don't think the lib64 issue changes the rationale for bundling language modules.

**Comment by andrewjpage** [comment]
Hi Folks,
I've updated it again:
squashed the commits into one
added resources
added the fasttree --with-double option
fixed the test check
fixed the style + audit.

Thanks for all the assistance,
Andrew

**Comment by tdsmith** [comment]
dendropy is struggling to learn its version:
```
==> python3 -c import setuptools... --no-user-cfg install --prefix=/usr/local/Cellar/gubbins/1.4.0/libexec/vendor --single-version-externally-managed --record=installed.txt
Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "setup.py", line 30, in <module>
    sys.stderr.write("-setup.py: {}\n".format(description()))
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/__init__.py", line 145, in description
    fields["DendroPy version"] = name()
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/__init__.py", line 122, in name
    return "{} {}{}".format(__project__, __version__, revision_description())
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/__init__.py", line 115, in revision_description
    if __revision__.is_available:
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/utility/vcsinfo.py", line 122, in is_available
    self.update()
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/utility/vcsinfo.py", line 128, in update
    if not self.repo_path or not self._vcs_available():
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/utility/vcsinfo.py", line 164, in _vcs_available
    retcode, stdout, stderr = self._run_vcs("--version")
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/utility/vcsinfo.py", line 157, in _run_vcs
    stdout, stderr = processio.communicate(p)
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/utility/processio.py", line 46, in communicate
    stdout = textprocessing.bytes_to_text(stdout)
  File "/private/tmp/gubbins--dendropy20150706-65853-1m1wa46/DendroPy-4.0.2/dendropy/utility/textprocessing.py", line 44, in bytes_to_text
    s = codecs.decode(s, ENCODING)
TypeError: must be str, not None
```

**Comment by tseemann** [comment]
@andrewjpage - my prediction is you are now at the stage of preferring to poke your eye with a fork?

**Comment by andrewjpage** [comment]
Yes @tseemann I'm nearly at that stage alright!

**Comment by andrewjpage** [comment]
Right so I've tested it on Ubuntu Trusty & Precise, and on Mountain Lion.  I havent come across the dendropy error I'm afraid but hopefully other changes I've made have fixed it...

**Comment by andrewjpage** [comment]
Hi @tdsmith and @tseemann ,
I was wondering if this pull request could be looked at again? Based on feedback from @tdsmith on other pull requests, I've update this one to bring it into line.
Regards,
Andrew

**Comment by tdsmith** [comment]
It seems wrong that biopython can't find numpy; I'll take a look.

**Comment by andrewjpage** [comment]
My guess is that numpy was already installed but without the python3 flag
set?

**Comment by tdsmith** [comment]
Here's the real error:
```
bash-3.2$ python3
Python 3.4.3 (default, Jul 13 2015, 12:18:23) 
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/numpy/__init__.py", line 170, in <module>
    from . import add_newdocs
  File "/usr/local/lib/python2.7/site-packages/numpy/add_newdocs.py", line 13, in <module>
    from numpy.lib import add_newdoc
  File "/usr/local/lib/python2.7/site-packages/numpy/lib/__init__.py", line 8, in <module>
    from .type_check import *
  File "/usr/local/lib/python2.7/site-packages/numpy/lib/type_check.py", line 11, in <module>
    import numpy.core.numeric as _nx
  File "/usr/local/lib/python2.7/site-packages/numpy/core/__init__.py", line 6, in <module>
    from . import multiarray
ImportError: dlopen(/usr/local/lib/python2.7/site-packages/numpy/core/multiarray.so, 2): Symbol not found: _PyBuffer_Type
  Referenced from: /usr/local/lib/python2.7/site-packages/numpy/core/multiarray.so
  Expected in: flat namespace
 in /usr/local/lib/python2.7/site-packages/numpy/core/multiarray.so
```

Something is setting PYTHONPATH inappropriately:
```
bash-3.2$ echo $PYTHONPATH
/usr/local/Cellar/gubbins/1.4.1/libexec/vendor/lib/python3.4/site-packages:/usr/local/Cellar/gubbins/1.4.1/libexec/lib/python3.4/site-packages:/usr/local/lib/python2.7/site-packages
```

**Comment by tdsmith** [comment]
Oh, I see. Pillow depends_on :python, so a dependency on system python2 is being pulled in, and we set PYTHONPATH for system python2. Should be fixed by https://github.com/Homebrew/homebrew-python/pull/238.

**Comment by tdsmith** [comment]
Bot still doesn't like dendropy.

```
==> python3 -c import setuptools... --no-user-cfg install --prefix=/usr/local/Cellar/gubbins/1.4.1/libexec/vendor --single-version-externally-managed --record=installed.txt
Traceback (most recent call last):
  File "<string>", line 4, in <module>
  File "setup.py", line 30, in <module>
    sys.stderr.write("-setup.py: {}\n".format(description()))
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/__init__.py", line 145, in description
    fields["DendroPy version"] = name()
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/__init__.py", line 122, in name
    return "{} {}{}".format(__project__, __version__, revision_description())
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/__init__.py", line 115, in revision_description
    if __revision__.is_available:
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/utility/vcsinfo.py", line 122, in is_available
    self.update()
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/utility/vcsinfo.py", line 128, in update
    if not self.repo_path or not self._vcs_available():
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/utility/vcsinfo.py", line 164, in _vcs_available
    retcode, stdout, stderr = self._run_vcs("--version")
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/utility/vcsinfo.py", line 157, in _run_vcs
    stdout, stderr = processio.communicate(p)
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/utility/processio.py", line 46, in communicate
    stdout = textprocessing.bytes_to_text(stdout)
  File "/private/tmp/gubbins--dendropy20150714-67986-1hfqyw0/DendroPy-4.0.2/dendropy/utility/textprocessing.py", line 44, in bytes_to_text
    s = codecs.decode(s, ENCODING)
TypeError: must be str, not None
```

**Comment by tdsmith** [comment]
Looks like that's a dendropy bug, fixed here: https://github.com/jeetsukumaran/DendroPy/commit/c9c1c097f201f2ab2962ac450938c430b3d3dd2f -- try setting `ENV["LANG"] = "en_US.UTF-8"` at the top of `install`.

**Comment by andrewjpage** [comment]
Thanks for taking the time to look at it in such detail. I owe you a few pints at this stage. I've made the change you suggested.

**Comment by tdsmith** [comment]
Houston, we have liftoff! Thanks for your patience.

**Comment by andrewjpage** [comment]
Thanks a million for all the work on getting it sorted!

**Pull request closed without merging**